Module porting and miscellaneous typing fixes

tristan957 opened the pull request.

tristan957 committed:
Add some typing to the dlang module

eli-schwartz commented on mesonbuild/modules/modtest.py:
```diff
@@ -24,21 +24,23 @@
+    def print_hello(self, state: ModuleState, args: T.List[TYPE_var], kwargs: TYPE_kwargs) -> ModuleReturnValue:
         print('Hello from a Meson module')
 
+        return ModuleReturnValue(None, [])
```
Why?

eli-schwartz commented on mesonbuild/modules/sourceset.py:
```diff
@@ -197,8 +197,6 @@ def add_all_method(self, state: ModuleState, args: T.Tuple[T.List[SourceSetImpl]
         keys, dependencies = self.check_conditions(when)
         for s in if_true:
-            if not isinstance(s, SourceSetImpl):
-                raise InvalidCode('Arguments to \'add_all\' after the first must be source sets')
```
Indeed unnecessary, this code predated typed_pos_args/typed_kwargs.

tristan957 replied:
Incorporated comment into commit message.

eli-schwartz commented on mesonbuild/modules/simd.py:
```diff
@@ -13,76 +13,119 @@
 
-from .. import mesonlib, compilers, mlog
+import typing as T
+from mesonbuild.interpreter.type_checking import STATIC_LIB_KWS, SOURCES_KW
+from mesonbuild.interpreterbase.decorators import KwargInfo, typed_pos_args, typed_kwargs
+
+from .. import mesonlib, mlog
 from .. import build
+from ..compilers import Compiler
```
The order of the imports you're adding seems all over the place... and why are some of them `from mesonbuild.`

eli-schwartz replied:
In general, this commit seems to have leaned very heavily into avoiding cases where git can recognize previous content in the diff.

eli-schwartz commented on mesonbuild/modules/simd.py:
```diff
@@ -13,76 +13,119 @@
-            raise mesonlib.MesonException('Argument must be a string.')
-        if 'compiler' not in kwargs:
-            raise mesonlib.MesonException('Must specify compiler keyword')
-        if 'sources' in kwargs:
-            raise mesonlib.MesonException('SIMD module does not support the "sources" keyword')
```
This check appears to have been dropped.

eli-schwartz commented on mesonbuild/modules/simd.py:
```diff
@@ -13,76 +13,119 @@
+
+    @typed_pos_args('simd.check', str)
+    @typed_kwargs('simd.check',
+                  KwargInfo('compiler', (Compiler,), required=True),
```
If it's just one type, don't pass it as a tuple.

eli-schwartz commented on mesonbuild/modules/icestorm.py:
```diff
@@ -34,17 +34,19 @@ class ProjectKwargs(TypedDict):
         constraint_file: T.Union[mesonlib.FileOrString, build.GeneratedTypes]
 
-class IceStormModule(ExtensionModule):
+class IceStormModule(NewExtensionModule):
```
Do not update this to the new extension module API. The purpose of the new extension module API is specifically, solely, exclusively, to drop access to self.interpreter... which you have added here in the `__init__` method for IceStormModule instead?

It is correct to share implementation of assigning `self.interpreter`.

eli-schwartz commented on mesonbuild/modules/i18n.py:
```diff
@@ -285,7 +285,7 @@ def gettext(self, state: 'ModuleState', args: T.Tuple[str], kwargs: 'Gettext') -
         assert isinstance(install_dir, str), 'for mypy'
         if not languages:
-            languages = read_linguas(path.join(state.environment.source_dir, state.subdir))
+            languages = read_linguas(path.join(state.environment.get_source_dir(), state.subdir))
```
This is a typing fix? :eyes:

tristan957 committed:
Remove unused import in internaltests.py

eli-schwartz commented on mesonbuild/ast/interpreter.py:
```diff
@@ -273,15 +273,12 @@ def reduce_arguments(
                 key_resolver: T.Callable[[mparser.BaseNode], str] = default_resolve_key,
                 duplicate_key_error: T.Optional[str] = None,
             ) -> T.Tuple[T.List[TYPE_nvar], TYPE_nkwargs]:
-        if isinstance(args, ArgumentNode):
```
You're going to have to work quite hard to convince me that removing this isinstance check is "removing a typing comment". ;)

eli-schwartz commented on mesonbuild/backend/backends.py:
```diff
@@ -1459,20 +1459,14 @@ def get_custom_target_sources(self, target: build.CustomTarget) -> T.List[str]:
+    def get_custom_target_depend_files(self, target: T.Union[build.CustomTarget, build.GeneratedList], absolute_paths: bool = False) -> T.List[str]:
         deps: T.List[str] = []
         for i in target.depend_files:
-            if isinstance(i, mesonlib.File):
```
Deleting checks for depend_files that are strings, is a typing fix?

tristan957 replied:
depend_files is marked as a List of strings, but that is apparently wrong

eli-schwartz replied:
Note that issues like this are why I don't trust logic changes to be inside of type annotation fixes unless the commit message explains why that individual logic change is correct and the previous version was wrong. Type annotations are there to hint at what is possible, but they are not authoritative (at least not for a codebase that isn't fully typed yet).

tristan957 replied:
Some day...

eli-schwartz commented on mesonbuild/backend/backends.py:
```diff
@@ -1829,8 +1823,6 @@ def generate_header_install(self, d: InstallData) -> None:
 
             for f in h.get_sources():
-                if not isinstance(f, File):
-                    raise MesonException(f'Invalid header type {f!r} can\'t be installed')
```
Note to self: this dead code was made impossible in 251dff56fb074cd718da81c21e104cd26d897869

tristan957 replied:
Incorportated into commit message.

dcbaker commented on mesonbuild/modules/simd.py:
```diff
@@ -13,76 +13,119 @@
+            my_kwargs[lang_args_key] = all_lang_args
+
+            # TODO: Seems like a layering violation
+            lib = self.interpreter.func_static_lib(state.current_node, [my_name], my_kwargs)
```
I had looked at this before, I think it makes more sense to use:
```python
lib = self.interpreter.build_target(state.current_node, (my_name, ), my_kwargs, build.StaticLibrary)
```
which avoids running back through the validators.

dcbaker commented on mesonbuild/compilers/mixins/arm.py:
```diff
@@ -34,43 +34,43 @@
+}
 
-armclang_optimization_args = {
+armclang_optimization_args: T.Dict[str, T.List[str]] = {
```
while you're here, can we use `T.Mapping[str, T.List[str]]`? This will have mypy warn us if we try to manipulate what is supposed to be module constant data.

dcbaker commented on mesonbuild/modules/dlang.py:
```diff
@@ -56,7 +57,7 @@ def _init_dub(self, state):
 
     @typed_pos_args('dlang.generate_dub_file', str, str)
-    def generate_dub_file(self, state, args, kwargs):
+    def generate_dub_file(self, state: ModuleState, args, kwargs):
```
since you're here, `args: T.Tuple[str, str]`

dcbaker commented on mesonbuild/build.py:
```diff
@@ -1517,10 +1517,10 @@ def add_include_dirs(self, args: T.Sequence['IncludeDirs'], set_is_system: T.Opt
         self.include_dirs += ids
 
-    def add_compiler_args(self, language: str, args: T.List['FileOrString']) -> None:
+    def add_compiler_args(self, language: str, args: T.List[str]) -> None:
```
This is actually allowed to be `File | str`, which seems weird but it's true.

dcbaker commented on mesonbuild/build.py:
```diff
@@ -1409,7 +1409,7 @@ def link(self, targets):
             self.link_targets.append(t)
 
-    def link_whole(self, targets, promoted: bool = False):
+    def link_whole(self, targets: T.Iterable[BuildTarget], promoted: bool = False):
```
`-> None:` while you're here

tristan957 commented on mesonbuild/modules/simd.py:
```diff
@@ -13,76 +13,119 @@
-            if args is None:
-                mlog.log('Compiler supports %s:' % iset, mlog.red('NO'))
+
+            sources = kwargs.get(iset, [])
```
Is this not indexing a dict with a key that might not exist?

dcbaker replied:
if it's in the `typed_kwargs` it's guaranteed to exist

tristan957 replied:
mesonbuild/modules/simd.py:98:30: error: TypedDict key must be a string literal; expected one of ("compiler", "mmx", "sse", "sse2", "sse3", ...)  [literal-required]

tristan957 committed:
Help type checkers out with "potentially" unbound variable

Although we know that this code can't execute if the language isn't one
of c, cpp, or fortran, some type checkers aren't smart enough to realize
that.

tristan957 committed:
Fix bad method overrides

Variable names must match.

tristan957 committed:
Use consistent type for the wanted kwarg

The typing for this kwarg was a tad inconsistent. version_compare() is
marked as str | List[str]. Use None | str | List[str] everywhere prior
to calling into version_compare() code paths.

tristan957 committed:
Use more explicit types

Some return values were missing, but overwhelmingly this commit just
adds explicit type annotations for variables. This assists type checkers
and can help the programmer know if that added an int to what was an
implicit list of strings, for instance.

tristan957 committed:
Use an underscore for variables that aren't used

tristan957 committed:
Remove some type quoting

tristan957 committed:
Use more correct types

Some types weren't marked as optional, but could be. Etc.

tristan957 committed:
Move typing outside of conditional in mpi_factory()

This assists type checkers such that cls is always the same type versus
a union otherwise.

tristan957 committed:
Clean up some boolean logic in wrap.py

This is the same thing as asking if depth is a key in self.wrap.values.

tristan957 committed:
Switch some Dict usage to Mapping

Mapping is immutable, which mypy can help us with.

tristan957 committed:
Use consistent type for extra_message in decorators.py

tristan957 committed:
Use typed_pos_args/typed_kwargs for the CMake module

tristan957 committed:
Fix boolean logic around whether or not a CMake subproject is found

tristan957 committed:
Add mesonbuild/modules/cmake.py to run_mypy.py

tristan957 committed:
Add mesonbuild/backend/nonebackend.py to run_mypy.py

tristan957 committed:
Port the i18n module to NewExtensionModule

eli-schwartz commented on mesonbuild/wrap/wrap.py:
```diff
@@ -643,11 +643,13 @@ def get_netrc_credentials(self, netloc: str) -> T.Optional[T.Tuple[str, str]]:
+            return None
+
+        _, account, password = info
 
-        return login, password
+        return account, password
```
You say we don't need to handle account being None "since account can never be None", but why? Particularly since the functionality logic here is "if account was present in the netrc file, prefer it instead of login".

You raise an interesting question in this commit, but the answer to it is, I think, that typing isn't everything, and that changing functional logic because typing says to is not necessarily correct, ***even if*** the change is correct according to typing.

... Although this change is NOT correct according to typing, if you take a close look at `typeshed/stdlib/netrc.pyi`.

eli-schwartz commented on mesonbuild/modules/python3.py:
```diff
@@ -38,35 +40,37 @@ class Python3Module(ExtensionModule):
             'sysconfig_path': self.sysconfig_path,
         })
 
-    @permittedKwargs(known_shmod_kwargs - {'name_prefix', 'name_suffix'})
```
What???

Also just want to reiterate that this functional change cannot be validly described by the commit message "Port to the new module API and add it to mypy."

tristan957 replied:
Added this back and folded most of the commit into other commits which more match what the changes actually are.

eli-schwartz commented on mesonbuild/modules/__init__.py:
```diff
@@ -87,15 +87,15 @@ def get_include_args(self, include_dirs: T.Iterable[T.Union[str, build.IncludeDi
                      required: bool = True,
                      version_func: T.Optional[T.Callable[[T.Union[ExternalProgram, build.Executable, OverrideProgram]], str]] = None,
-                     wanted: T.Optional[str] = None, silent: bool = False,
+                     wanted: str = '', silent: bool = False,
```
Hmm, interesting question which one should be fixed :P since interpreter.py is typed up as `wanted: T.Union[str, T.List[str]] = '',` but checks via `if wanted:`. So do we want to pass an empty string around, or None? Both behave the same as far as find_program_impl() is concerned.

None is a true singleton...

eli-schwartz commented on mesonbuild/arglist.py:
```diff
@@ -264,8 +264,8 @@ def extend_direct(self, iterable: T.Iterable[str]) -> None:
 
     def extend_preserving_lflags(self, iterable: T.Iterable[str]) -> None:
-        normal_flags = []
-        lflags = []
+        normal_flags: T.List[str] = []
+        lflags: T.List[str] = []
```
> overwhelmingly this commit just adds explicit type annotations for variables. This assists type checkers and can help the programmer know if that added an int to what was an implicit list of strings, for instance.

mypy actually infers this as T.List[str] since the first action performed on it is to append something of type str. So appending an int would be a type error, but if the first append was for an int, then appending a str would be a type error.

e.g. if you try adding `.append(1)` at the end of the function, it throws:
> Argument 1 to "append" of "list" has incompatible type "int"; expected "str"  [arg-type]

If it cannot figure out how to properly type narrow on the first usage, then does demand you explicitly annotate it:
> Need type annotation for "mylist" (hint: "mylist: List[<type>] = ...")  [var-annotated]

tristan957 replied:
This commit is really an aid for VSCode users. Though mypy seems to be happy with most of these types, pylance/pyright need a little aid, and I find the readability a little bit easier. Is there anything actionable you want me to do here?

tristan957 commented:
Test failure in Ubuntu Rolling are pretty suspicious. I can't recreate them locally. It is in a very stupid code path too. Mixing booleans and strings.... :(

eli-schwartz commented on mesonbuild/envconfig.py:
```diff
@@ -400,9 +400,6 @@ def __init__(
         self.binaries: T.Dict[str, T.List[str]] = {}
         if binaries:
             for name, command in binaries.items():
-                if not isinstance(command, (list, str)):
-                    raise mesonlib.MesonException(
-                        f'Invalid type {command!r} for entry {name!r} in cross file')
```
Drop this commit altogether. It isn't unused, no matter what the typing says. For one simple reason:

The typing says how this function is supposed to be used. It doesn't say that external users are *unable* to use it incorrectly. And this function's input is... a machine file specified on the command line. This code right here is the exclusive code for validating user input.

tristan957 replied:
Done

eli-schwartz commented on mesonbuild/interpreterbase/decorators.py:
```diff
@@ -786,7 +786,7 @@ def feature_check_class(self) -> T.Type[FeatureCheckBase]:
 
     def __init__(self, feature_name: str, feature_version: str,
-                 kwargs: T.List[str], extra_message: T.Optional[str] = None):
+                 kwargs: T.List[str], extra_message: str = ''):
```
This isn't the only place that it is Optional[str], but I suspect the better solution is to make the one instance be initialized as None rather than changing the two instances to initialize as a blank string.

tristan957 replied:
I agree. Fixed.

tristan957 committed:
Add the mesonbuild/modules/python3.py to run_mypy.py

tristan957 commented on mesonbuild/wrap/wrap.py:
```diff
@@ -643,11 +643,13 @@ def get_netrc_credentials(self, netloc: str) -> T.Optional[T.Tuple[str, str]]:
-            login = account
+        info = self.netrc.authenticators(netloc)
+        if info is None:
+            return None
```
Commit dropped.

eli-schwartz commented:
I cherry-picked some of these changes which I've reviewed: 179355bf18...ed9d03a088

tristan957 commented:
Cool. I will rebase after getting the simd module stuff finished and see what else we can do here
